weld/codegen/llvm2/cmp.rs

```diff
@@ -0,0 +1,351 @@
+                    LLVMExtAddAttrsOnParameter(self.context, memcmp, &[ReadOnly, NoCapture], 1);
+                }
+                let ref mut args = [left_data, right_data, bytes];
+                let memcmp_result = self.intrinsics.call(builder, name, args)?;
+
+                LLVMBuildRet(builder, memcmp_result);
+            }
+            Vector(ref elem) => {
+                // Compare vectors with a loop. Check size like before.
+                let loop_block = LLVMAppendBasicBlockInContext(self.context(), function,  c_str!(""));
+                let done_block = LLVMAppendBasicBlockInContext(self.context, function, c_str!(""));
+
+                let left_vector = self.load(builder, left)?;
+                let right_vector = self.load(builder, right)?;
+                let left_size = self.gen_size(builder, ty, left_vector)?;
+                let right_size = self.gen_size(builder, ty, right_vector)?;
+                let min_size = gen_binop(builder, Min, left_size, right_size, &Scalar(I64))?;
+
+                // Check if there are any elements to loop over.
+                let check = LLVMBuildICmp(builder, LLVMIntNE, min_size, self.i64(0), c_str!(""));
+
+                LLVMBuildCondBr(builder, check, loop_block, done_block);
+                LLVMPositionBuilderAtEnd(builder, loop_block);
+
+                // Index variable.
+                let phi_i = LLVMBuildPhi(builder, self.i64_type(), c_str!(""));
+
+                let func = self.gen_cmp_fn(elem)?;
+                let left_value = self.gen_at(builder, ty, left_vector, phi_i)?;
+                let right_value = self.gen_at(builder, ty, right_vector, phi_i)?;
+                let mut args = [left_value, right_value];
+                let cmp_result = LLVMBuildCall(builder, func, args.as_mut_ptr(), args.len() as u32, c_str!(""));
+                let neq = LLVMBuildICmp(builder, LLVMIntNE, cmp_result, self.i32(0), c_str!(""));
+
+                let updated_i = LLVMBuildNSWAdd(builder, phi_i, self.i64(1), c_str!(""));
+                let check1 = LLVMBuildICmp(builder, LLVMIntEQ, updated_i, left_size, c_str!(""));
+                let check2 = LLVMBuildOr(builder, check1, neq, c_str!(""));
+                // End loop if i == size || left != right
+                LLVMBuildCondBr(builder, check2, done_block, loop_block);
+
+                let mut blocks = [entry_block, loop_block];
+                let mut values = [self.i64(0), updated_i];
+                LLVMAddIncoming(phi_i, values.as_mut_ptr(), blocks.as_mut_ptr(), values.len() as u32);
+
+                // Finish block - set result and compute number of consumed bits.
+                LLVMPositionBuilderAtEnd(builder, done_block);
+                let result = LLVMBuildPhi(builder, self.i32_type(), c_str!(""));
+                let mut blocks = [entry_block, loop_block];
+                let mut values = [self.i32(1), cmp_result];
```

Comment:
I'd change `self.i32(1)` to `self.i32(0)`, since in the memcmp version, if there are no bytes to compare, `memcmp` will return 0 (and then the size check happens).